util/src/main/java/io/kubernetes/client/util/generic/GenericKubernetesApi.java

```diff
@@ -789,30 +1277,161 @@ KubernetesApiResponse<DataType> getKubernetesApiResponse(
     return new KubernetesApiResponse<>(gson.fromJson(element, dataClass));
   }
 
+  private Call prepareCall(CallBuilder callBuilder) throws ApiException {
+    Call call = callBuilder.build();
+    return tweakCallForCoreV1Group(call);
+  }
+
+  private void checkForIOException(ApiException e) {
+    if (e.getCause() instanceof IOException) {
+      throw new IllegalStateException(e.getCause()); // make this a checked exception?
+    }
+  }
+
+  private <DataType extends KubernetesType>
+      KubernetesApiResponse<DataType> responseFromApiException(
+          ApiClient apiClient, ApiException e) {
+    checkForIOException(e);
+    final V1Status status;
+    try {
+      status = apiClient.getJSON().deserialize(e.getResponseBody(), V1Status.class);
+    } catch (JsonSyntaxException jsonEx) {
+      // craft a status object
+      return new KubernetesApiResponse<>(
+          new V1Status().code(e.getCode()).message(e.getResponseBody()), e.getCode());
+    }
+    if (null == status) { // the response body can be something unexpected sometimes...
+      // this line should never reach?
+      throw new RuntimeException(e);
+    }
+    return new KubernetesApiResponse<>(status, e.getCode());
+  }
+
   private <DataType extends KubernetesType> KubernetesApiResponse<DataType> executeCall(
       ApiClient apiClient, Class<DataType> dataClass, CallBuilder callBuilder) {
     try {
-      Call call = callBuilder.build();
-      call = tweakCallForCoreV1Group(call);
+      Call call = prepareCall(callBuilder);
       JsonElement element = apiClient.<JsonElement>execute(call, JsonElement.class).getData();
       return getKubernetesApiResponse(dataClass, element, apiClient.getJSON().getGson());
     } catch (ApiException e) {
-      if (e.getCause() instanceof IOException) {
-        throw new IllegalStateException(e.getCause()); // make this a checked exception?
-      }
-      final V1Status status;
-      try {
-        status = apiClient.getJSON().deserialize(e.getResponseBody(), V1Status.class);
-      } catch (JsonSyntaxException jsonEx) {
-        // craft a status object
-        return new KubernetesApiResponse<>(
-            new V1Status().code(e.getCode()).message(e.getResponseBody()), e.getCode());
-      }
-      if (null == status) { // the response body can be something unexpected sometimes..
-        // this line should never reach?
-        throw new RuntimeException(e);
-      }
-      return new KubernetesApiResponse<>(status, e.getCode());
+      return responseFromApiException(apiClient, e);
+    }
+  }
+
+  // For unit-test
+  interface CallAdapter extends UnaryOperator<Call> {}
+
+  private <DataType extends KubernetesType> Call prepareCall(
+      CallBuilder callBuilder, @Nonnull Consumer<KubernetesApiResponse<DataType>> callback)
+      throws ApiException {
+    Call call = prepareCall(callBuilder);
+
+    // For unit-test
+    if (callback instanceof CallAdapter) {
```

Comment:
i always have a complex feeling about having these unit-test-only code in the main path. as for this `CallAdapter` interface, the mere usage i found is a `TestCallback` which is for retrieving response data from the client, and if i understand correctly they can be replaced by `Future#get()`, so we dont `CallAdapter` here, is that correct?

Reply:
The unit-test code (GenericKubernetesApiForCoreApiTest, line 552) needs to modify the Call before it is executed. I did not want to expose the Call type (or any of the OkHttp types) through the GenericKubernetesApi API. 

Because the Call needs to be modified before it is executed or enqueued, this call can't be replaced with Future#get().

If you would prefer, I could expose a supported API to modify the Call before it is used. I've had good success with unit-test specific code that is package private.

Reply:
> The unit-test code (GenericKubernetesApiForCoreApiTest, line 552) needs to modify the Call before it is executed.

oh i missed that unit-test. ideally it would be nice if there're other workaround to remove those test-only code. but we can continue to refactor them in follow-ups when we have a better idea.